server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java

```diff
@@ -127,22 +132,22 @@ public static List<NodeAllocationResult> sortNodeDecisions(List<NodeAllocationRe
     }
 
     /**
-     * Generates X-Content for the node-level decisions, creating the outer "node_decisions" object
+     * Generates X-Content chunks for the node-level decisions, creating the outer "node_allocation_decisions" object
      * in which they are serialized.
      */
-    public static XContentBuilder nodeDecisionsToXContent(List<NodeAllocationResult> nodeDecisions, XContentBuilder builder, Params params)
-        throws IOException {
-
-        if (nodeDecisions != null && nodeDecisions.isEmpty() == false) {
-            builder.startArray("node_allocation_decisions");
-            {
-                for (NodeAllocationResult explanation : nodeDecisions) {
-                    explanation.toXContent(builder, params);
-                }
-            }
-            builder.endArray();
+    public static Iterator<ToXContent> nodeDecisionsToXContentChunked(List<NodeAllocationResult> nodeDecisions) {
+        if (nodeDecisions == null || nodeDecisions.isEmpty()) {
+            return Collections.emptyIterator();
         }
-        return builder;
+
+        return Iterators.concat(
+            ChunkedToXContentHelper.startArray("node_allocation_decisions"),
+            Iterators.flatMap(
+                nodeDecisions.listIterator(),
+                explanation -> Iterators.single((builder, params) -> explanation.toXContent(builder, params))
+            ),
```

Comment:
`flatMap`/`single` is a no-op:

```suggestion
            nodeDecisions.iterator(),
```

Reply:
I don't think that'd prevent the `flatMap`/`single` case. That would forbid returning an empty iterator for one of the items over which we're iterating, but that's a legitimate thing to do. See e.g. here:

https://github.com/elastic/elasticsearch/blob/19257125b19ff05a6b014f37693f9417ca579626/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L1453-L1458

I don't have a great idea for avoiding this apart from catching it in reviews. I imagine we could write a checkstyle rule for it if it was very important to us, but the consequences of missing it are not terribly severe.